tests/Unit/NumericalAlgorithms/RootFinding/Test_NewtonRaphson.cpp

```diff
@@ -38,44 +37,53 @@ void test_simple() {
     return std::make_pair(2. - square(x), -2. * x);
   };
   const auto root_from_lambda = RootFinder::newton_raphson(
-      func_and_deriv_lambda, guess, lower, upper, digits);
+      func_and_deriv_lambda, guess, lower, upper, residual_tolerance,
+      step_absolute_tolerance, step_relative_tolerance);
   // [double_newton_raphson_root_find]
   const auto root_from_free = RootFinder::newton_raphson(
-      func_and_deriv_free, guess, lower, upper, digits);
+      func_and_deriv_free, guess, lower, upper, residual_tolerance,
+      step_absolute_tolerance, step_relative_tolerance);
   const FuncAndDeriv func_and_deriv_functor{};
   const auto root_from_functor = RootFinder::newton_raphson(
-      func_and_deriv_functor, guess, lower, upper, digits);
-  CHECK(std::abs(root_from_lambda - correct) < 1.0 / std::pow(10, digits));
+      func_and_deriv_functor, guess, lower, upper, residual_tolerance,
+      step_absolute_tolerance, step_relative_tolerance);
+  CHECK(std::abs(root_from_lambda - correct) <= step_absolute_tolerance);
   CHECK(root_from_free == root_from_lambda);
   CHECK(root_from_free == root_from_functor);
 }
 
 void test_bounds() {
-  const size_t digits = 8;
+  const double residual_tolerance = 1.0e-14;
+  const double step_absolute_tolerance = 1.0e-8;
+  const double step_relative_tolerance = 0.0;
   const double guess = 1.5;
   double upper = 2.;
   double lower = sqrt(2.);
   const auto func_and_deriv_lambda = [](double x) {
     return std::make_pair(2. - square(x), -2. * x);
   };
 
-  auto root = RootFinder::newton_raphson(func_and_deriv_lambda, guess, lower,
-                                         upper, digits);
+  auto root = RootFinder::newton_raphson(
+      func_and_deriv_lambda, guess, lower, upper, residual_tolerance,
+      step_absolute_tolerance, step_relative_tolerance);
   const double correct = sqrt(2.);
 
-  CHECK(std::abs(root - correct) < 1.0 / std::pow(10, digits));
+  CHECK(std::abs(root - correct) < step_absolute_tolerance);
 
   lower = 0.;
   upper = sqrt(2.);
 
   root = RootFinder::newton_raphson(func_and_deriv_lambda, guess, lower, upper,
-                                    digits);
-  CHECK(std::abs(root - correct) < 1.0 / std::pow(10, digits));
+                                    residual_tolerance, step_absolute_tolerance,
+                                    step_relative_tolerance);
+  CHECK(std::abs(root - correct) < step_absolute_tolerance);
```

Comment:
how many iterations does this take?  If the root is at the lower bound, you exit on the first iteration, but there is not a similar check at the upper bound

Reply:
wait, sorry, the lower bound is only evaluated if you need to do a bracket, so not necessarily the first iteration;  maybe evaluate at the upper bound if it becomes part of the bracket?  @teukolsky do you have an opinion on when (if ever) to evaluate at the initial bounds?

Reply:
I'll check numbers tomorrow, but remember that for most of our uses I've set the residual tolerance to zero (because the boost one didn't have that option), so you have to get an exact zero from floating point math.  Of course, if the bounds are going to be the root you probably should have a non-zero residual tolerance so roundoff can't invalidate your bracket, so maybe it is useful to check them.

Reply:
I'm pretty hard core on this. The cost of the various checks is negligible compared with a function evaluation, and there's nothing worse than hard-to-debug runtime errors when the assumptions aren't satisfied. Here's what I would very much urge us to do, and not with asserts (the problems can arise in real runs):
1) Evaluate f1 and f2, the function values at the bounds (or pass them in).
2) Check the signs: if they're the same, the interval isn't a bound. This check should be done excluding zero as a function value.
3) Test f1 and f2 each for exactly 0. Return corresponding x if true.
4) Start the algorithm